### ISSUE TITLE
Suggestion: DNS proxy scripts

### ISSUE THREAD

**Issue by buffermet** [bettercap/bettercap#779]
I am writing this to consolidate https://github.com/bettercap/bettercap/pull/496 and https://github.com/bettercap/bettercap/pull/764#issuecomment-700779340 (and others).

It would be great to have the ability to have a more dynamic DNS spoofer by means of a new `dns.proxy` module.

We could then determine (with the use of JavaScript) when an IP does or doesn't get resolved, with which TTL, for who and by who.

It would also allow for a simpler implementation of [a captive portal module](https://github.com/bettercap/bettercap/issues/631).

**Comment by evilsocket** [comment]
as far as i can see #496 is not adding a new module but changing the default behaviour of an existing one, did i get it right? i'd merge right away a new module, but since there're a lot of users and software relying on that default behaviour i'm prone not to merge ... also the PR conflicts :/

**Comment by buffermet** [comment]
Yep, that's right.

I opened this issue to track this feature request. I think that if we add a new `dns.proxy` module, we can have a dynamic DNS proxy server that would make #496 and https://github.com/bettercap/bettercap/pull/764#issuecomment-700779340 obsolete.

**Comment by buffermet** [comment]
I've had a [dns.proxy module draft](https://github.com/buffermet/bettercap/tree/master/modules/dns_proxy) sitting in a forked repo that has been gathering dust for over 2 years. Time really flies...

There's also a [captive portal draft](https://github.com/buffermet/bettercap/blob/master/captiveportal.go). I should have a captive portal caplet folder somewhere that contains a separate version for macOS/Windows. The Linux version uses the packet proxy and is tunnel proof, while the other version relies on the DNS proxy and is not tunnel proof.

I have some free time to work on this project so I will aim to finalize testing this week.

**Comment by evilsocket** [comment]
@buffermet so good to see you here again my friend! time indeed flies :)

First: awesome! I'm certain that the new dns proxy module will make a lot of people happy :D I saw there are some conflicts, but given they're new modules it shouldn't be too hard to rebase for you. Then maybe PR? :D

So excited to see you active again <3

**Comment by evilsocket** [comment]
btw did you take a look at the (not really) new scripting mechanism? https://www.bettercap.org/usage/scripting/

I feel there's a lot more than it can be done with it rather than using only caplets ... wdyt?

**Comment by buffermet** [comment]
@evilsocket I think you're right on the money with those session scripts!

I ran into an issue where I had no other choice but to parse a JSON string of lease records each time there was a DNS or HTTP packet, which is obviously not something we wanna do on large networks

With a session script we can redefine the http.proxy scope using the blacklist, although the macOS/Windows version will only be able to mimic a captive portal (show sign-in notification and redirect DNS/HTTP traffic)

maybe I'll should just stick to 1 Linux version for now

**Comment by buffermet** [comment]
and it is very good to be back my friend, your write-up on the cups rce was very inspiring to say the least! :)

**Comment by evilsocket** [comment]
@buffermet at this point i think you know both the http and the dns modules better than i do :D 

I'm not sure I understand the core difference between the OSs here, can you explain please? As long as we use the [core abstraction around the firewalls](https://github.com/bettercap/bettercap/tree/master/firewall) we should be OS independent ... no? What am I missing? Can you ELI5? :'D

**Comment by evilsocket** [comment]
(thank you! glad you liked it! <3)

**Comment by evilsocket** [comment]
btw did you check / test zerogod? https://github.com/bettercap/bettercap/tree/master/modules/zerogod i think this will be the future, zeroconf is everywhere and i know this .... maybe call sometimes? :D

**Comment by buffermet** [comment]
@evilsocket I know the basics :D there are so many protocols over DNS (and HTTP) that I don't know, but the miekg/dns plugin supports them all

> btw did you check / test zerogod? https://github.com/bettercap/bettercap/tree/master/modules/zerogod i think this will be the future, zeroconf is everywhere and i know things .... maybe call sometimes? :D

I had some reading to do because I'm still not fully up to speed with mDNS and zeroconf... I thought mDNS was primarily used in services like bonjour for Apple ecosystem, but if I understand correctly, it's also used for serverless DHCP, local DNS and service streaming?

> I'm not sure I understand the core difference between the OSs here, can you explain please? As long as we use the [core abstraction around the firewalls](https://github.com/bettercap/bettercap/tree/master/firewall) we should be OS independent ... no? What am I missing? Can you ELI5? :'D

So if we want to have a module on macOS/Windows that can periodically block internet access for a specific client, how can we achieve this? We would need to use the packet.proxy module no? Since we can set the verdict to drop the packet if certain conditions aren't met.

The idea of the captive portal module is to block all outgoing upstream traffic and redirect DNS and HTTP requests to a server with a HTML login page, where the client can authenticate before gaining access to the internet. Is it possible without the packet.proxy?

**Comment by evilsocket** [comment]
> I had some reading to do because I'm still not fully up to speed with mDNS and zeroconf... I thought mDNS was primarily used in services like bonjour for Apple ecosystem, but if I understand correctly, it's also used for serverless DHCP, local DNS and service streaming?

Yeah man there's all sorts of services running on this stuff ... it's insane https://github.com/bettercap/bettercap/blob/master/modules/zerogod/zerogod_known_services.go

> So if we want to have a module on macOS/Windows that can periodically block internet access for a specific client, how can we achieve this? We would need to use the packet.proxy module no? Since we can set the verdict to drop the packet if certain conditions aren't met.
> The idea of the captive portal module is to block all outgoing upstream traffic and redirect DNS and HTTP requests to a server with a HTML login page, where the client can authenticate before gaining access to the internet. Is it possible without the packet.proxy?

As far as I know, more than blocking everything, captive portals REDIRECT everything. Basically any request to any website (in https too) gets hijacked and the login page returned. Any OS periodically check a specific page, for instance Apple uses http://captive.apple.com/hotspot-detect.html (notice the http instead of https) and if anything other than the expected response is returned, the OS assumes it's a captive portal and will show you the login page. 

So essentially it's just a matter (i believe) of http/https proxy redirecting anything to a login page, especially stuff like http://captive.apple.com/hotspot-detect.html

**Comment by buffermet** [comment]
> Yeah man there's all sorts of services running on this stuff ... it's insane https://github.com/bettercap/bettercap/blob/master/modules/zerogod/zerogod_known_services.go

🤯 it's a brave new (UDP) world

> So essentially it's just a matter (i believe) of http/https proxy redirecting anything to a login page, especially stuff like http://captive.apple.com/hotspot-detect.html

We definitely want to redirect all DNS and HTTP(S) traffic to the captive portal login server, but when clients are using a VPN (or more advanced UDP/ICMP) tunnel, they can bypass this system. So I'm thinking we want to have an industry standard captive portal for bettercap that also blocks the other traffic, but I think that the industry standard is Linux anyway (and maybe some airport utility on macOS).

Maybe it doesn't need the [rate limit mechanism](https://github.com/buffermet/bettercap/blob/5f5ae0455b25ef52f1c6ced151e261676f92e72c/captiveportal.go#L112) after all... I really should wrap my head around this entire module again xD

**Comment by evilsocket** [comment]
I see ... maybe we can work with the firewall to drop stuff instead of the packet proxy? that'd work on mac too

**Comment by buffermet** [comment]
I like the idea of supporting mac by using the firewall, but then we still have the problem of captive portal authentication.

I ended up going for a packet.proxy module since I don't need to create a special callback/IPC to handle user data. I want the option to set download limits, time limits, authentication requirements (cookies), etc. Still need to think of an elegant way to handle captive portal authentication...

I got the dns.proxy module working with a firewall redirect now, there are still issues, but at least there's some progress

![image](https://github.com/user-attachments/assets/3acaf89e-089c-4c67-9c0e-0c77b0e3c5e9)

**Comment by buffermet** [comment]
I really like the idea of using firewall but it's gonna require some refactoring I believe

We'd have the ability to change firewall rules using a session script for example?

**Comment by buffermet** [comment]
the only issue I can think of now is the authentication

let's assume for simplicity that the client needs to include a cookie header with a HTTP request in order to authenticate with the captive portal and gain internet access

would we be able to do this with a session script? I don't think the HTTP request session event includes the headers. This is where I ran into the problem of using environment variables and parsing JSON data upon receiving requests, but that doesn't seem like the best solution for this module

**Comment by evilsocket** [comment]
> We'd have the ability to change firewall rules using a session script for example?

Yes i was thinking that should definitely be added ... can you think / list a few primitives that would be needed for this usecase?

>  I don't think the HTTP request session event includes the headers.

i'd be lying if i said i remember XD but i guess we can augmnent what's sent to the scripts.

> maybe we can make a custom session event from the http proxy? that way it can communicate with the session script

yep that sounds like the most elegant solution

**Comment by buffermet** [comment]
> Yes i was thinking that should definitely be added ... can you think / list a few primitives that would be needed for this usecase?

honestly I think you've already solved every problem with the session scripts, since we can execute shell commands we can append a single redirect rule and then control a list of IPs using ipset (with sanity checks of course :D)

I think just like the onLoad() proxy script function, we should also have an onExit() or onQuit() function, that way we can clean up iptables more elegantly for example

> yep that sounds like the most elegant solution

I'll get started on that functionality right away, glad we are having this chat because I would have just skipped macOS

It didn't really occur to me that bettercap has become completely omnipotent with the session script capabilities :exploding_head: you have done amazing work bro <3

**Comment by evilsocket** [comment]
> honestly I think you've already solved every problem with the session scripts, since we can execute shell commands we can insert a single redirect rule and then control a list of IPs using ipset (with sanity checks of course :D)

Ahaha yeah that'd work :D

> I think just like the onLoad() proxy script function, we should also have an onExit() or onQuit() function, that way we can clean up iptables more elegantly for example

Yes this makes sense, will implement asap!

> you have done amazing work bro <3

*We did it* ... thank you bro <3

**Comment by buffermet** [comment]
maybe it's a better solution to have the session script write the data to user configuration files, that way the `http.proxy` module can be used to serve this data

**Comment by buffermet** [comment]
if otto supported setTimeout and setInterval functionality it would be easy to solve the problem of a time/usage based lease, maybe we could implement a fix for this?

**Comment by buffermet** [comment]
I think this is why I decided to go with the packet proxy :D but I still think the solution with macOS support is more elegant

**Comment by buffermet** [comment]
> if otto supported setTimeout and setInterval functionality it would be easy to solve the problem of a time/usage based lease expiration, maybe we could implement a fix for this?

there is [natto](https://github.com/robertkrimen/natto) which is a variant of otto that implements an event loop, although I have not tested this

**Comment by evilsocket** [comment]
> if otto supported setTimeout and setInterval functionality it would be easy to solve the problem of a time/usage based lease expiration, maybe we could implement a fix for this?

These functions are kind of tricky because they would require a GIL of some sort ... but we have the "tick" event at custom intervals ... so maybe we can just set a tick every second, increment a counter or whatever inside the callback and when the counter for a given client (maybe kept in a global dictionary?) reaches X, you know the lease expired ... or something like that ... wdyt?

Or maybe a new module, similar to tick, that you can configure to execute a callback after a certain time ...

**Comment by buffermet** [comment]
> These functions are kind of tricky because they would require a GIL of some sort ... but we have the "tick" event at custom intervals ... so maybe we can just set a tick every second, increment a counter or whatever inside the callback and when the counter for a given client (maybe kept in a global dictionary?) reaches X, you know the lease expired ... or something like that ... wdyt?
> 
> Or maybe a new module, similar to tick, that you can configure to execute a callback after a certain time ...

been thinking about this, maybe rather than implementing a new module, we could change the ticker module to allow multiple tickers?

the ticker module logic is already perfect for this since it starts with a delay before executing the `ticker.commands` (rather than executing before the delay), so if we can create multiple tickers, we can make javascript bindings for the `setTimeout` and `setInterval` functions

**Comment by evilsocket** [comment]
> been thinking about this, maybe rather than implementing a new module, we could change the ticker module to allow multiple tickers?

NAMED TICKERS!!!! FUCKING GENIOUS

Yes! I like it a lot! I can change the ticker related commands to accept an optional ticker name (optional so we keep backwards compatibility), if the name is specified then a custom event ( ticker.custom-name-here ) will be triggered, otherwise the default one! :D This can be useful in so many other ways dude.

I just started a new job so i'm not gonna be able to allocate a lot of time to this, but should be able to work on it on the weekend :D

**Comment by buffermet** [comment]
> > been thinking about this, maybe rather than implementing a new module, we could change the ticker module to allow multiple tickers?
> 
> NAMED TICKERS!!!! FUCKING GENIOUS
> 
> Yes! I like it a lot! I can change the ticker related commands to accept an optional ticker name (optional so we keep backwards compatibility), if the name is specified then a custom event ( ticker.custom-name-here ) will be triggered, otherwise the default one! :D This can be useful in so many other ways dude.
> 
> I just started a new job so i'm not gonna be able to allocate a lot of time to this, but should be able to work on it on the weekend :D

Ah yes!! That's a very clean solution, this is gonna be good :D

It turns out this dns.proxy module is more complex than I initially thought but it's coming together, hopefully PR this week :crossed_fingers:

**Comment by buffermet** [comment]
@evilsocket you're a legend!! it works like a charm :D

Is it an idea to allow the command field to be empty in `ticker.create <name> <period> <command>`? That way we can simply listen for the session event without writing a newline to stdOut on each tick? Or is there a different way we can do this with a shell command?

I made a polyfill for the setInterval, setTimeout, clearInterval and clearTimeout functions, but I think we're still missing the option to remove session event listeners. If we have a function like `removeEventListener('ticker.1', myTimeoutFunc)` then we won't get a buildup of unused session event listeners.

And there's another idea for a module that I can't wait to show you, but first I want to get all this out of the way :D

Also I'd be happy to implement these features, I'll make some new issues so we can self assign stuff :)

**Comment by buffermet** [comment]
I wrote earlier:

> but there are some scenarios where we want to relay some information from the session script environment to the HTTP/DNS proxy script environment

...but I just remember that this can already be achieved with the `onCommand(cmd)` function.

For example:

In the proxy script:
```js
function onCommand(cmd) {
    if (cmd.startsWith('http.proxy.custom.command ')) {
        // do stuff with the data
        return true
    }
}
```

In the session script:
```js
run('http.proxy.custom.command <data>')
```

Although maybe it's a good idea to add `env(varName)` and `env(varName, value)` functionality to the session script (if that isn't already included) so that we don't have to worry about escaping semicolons etc. as we could then pass the environment variable name in the `<data>` argument.
